src/adapter/src/client.rs

```diff
@@ -397,13 +400,79 @@ impl SessionClient {
+        // To support parameters we do an AST rewrites taking `$x` to
+        // `CAST($params[x-1].0 AS $params[x-1].1)`
+        fn inline_params(
+            params: &[(&str, &str)],
+            stmt: &mut mz_sql::ast::Statement<Raw>,
+        ) -> Result<(), anyhow::Error> {
+            struct ParamInliner<'a> {
+                params: &'a [(&'a str, &'a str)],
+                error: Option<anyhow::Error>,
+            }
+
+            impl<'ast, 'a> VisitMut<'ast, Raw> for ParamInliner<'a> {
+                fn visit_expr_mut(&mut self, node: &'ast mut Expr<Raw>) {
+                    if let Expr::Parameter(i) = node {
+                        let i = *i;
+                        // 1-indexed params make this awk
+                        let vals = match i {
+                            i if i == 0 => None,
+                            i => self.params.get(i - 1),
+                        };
+
+                        let (val, typ) = match vals {
+                            Some(v) => v,
+                            None => {
+                                self.error = Some(anyhow!("there is no parameter ${}", i));
+                                return;
+                            }
+                        };
+
+                        match mz_sql_parser::parser::parse_expr(&format!(
+                            "CAST ('{}' AS {})",
+                            display::escape_single_quote_string(val),
+                            typ
+                        )) {
+                            Ok(cast_expr) => {
+                                *node = cast_expr;
+                                assert!(
+                                    matches!(node, Expr::Cast { .. }),
+                                    "must convert to Expr::Cast"
+                                );
+                            }
+                            Err(_) => {
+                                self.error = Some(anyhow!(
+                                    "invalid parameter: CAST ('<val>' as <type>) must be valid expression"
+                                ));
+                            }
+                        }
+                    }
+                    visit_mut::visit_expr_mut(self, node);
+                }
+            }
+
+            let mut inliner = ParamInliner {
+                params,
+                error: None,
+            };
+            inliner.visit_statement_mut(stmt);
+            match inliner.error {
+                Some(err) => Err(err),
+                None => Ok(()),
+            }
+        }
+
         let stmts =
             mz_sql::parse::parse(&stmts).map_err(|e| AdapterError::Unstructured(e.into()))?;
         let num_stmts = stmts.len();
         const EMPTY_PORTAL: &str = "";
         let mut results = vec![];
-        for stmt in stmts {
+        for mut stmt in stmts {
+            inline_params(&params, &mut stmt).map_err(AdapterError::Unstructured)?;
```

Comment:
We should use the actual param stuff instead of inlining with strings. Can you use the same logic that the extended protocol does? The `parse`, `bind`, `execute` fns in protocol.rs have all the logic needed to do this correctly.

Reply:
This is how PG handles parameters in the simple protocol. Using portals, etc. turns this into some hybrid version of the simple and extended protocols, and adds more layers of complexity. Is there a concern with this implementation?

Reply:
PG doesn't handle parameters in the simple protocol. Where did you see that?

Reply:
Ah, I had misunderstood the PG documentation w/r/t parameter in the Simple Query API. Bad point.

However, I still don't think inlining the extended protocol approach is right here, unless we want to move all of the requests that use parameters to instead be processed that way?

Some of the things that threw me off of that approach:
- `StateMachine::parse` takes strings and only permits single statements to be executed; this means we'd have to take the input to Vec<statements>, and then take each of those individual statements back to strings to be parsed. That smells fishy to me.
-  `StateMachine::parse` requires inputting the `param_oid`s, but `simple_execute` doesn't have catalog access, so we've have to perform some complex queries to try and infer the correct OID for the type name--name resolution is complex and trying to manifest that in the HTTP API seems very challenging.

Doing an AST rewrite accomplishes the same thing simply and safely, so I'm very apprehensive about trying to bring the `parse`/`bind`/`execute`  code into this. If you'd like to spend some time with me showing me how you'd approach it, though, very glad to set up some time.

Reply:
For posterity, I under misunderstood Matt's suggestion; he was not talking about using those functions but instead using the logic they contain.

@mjibson can you summarize why you think this approach is superior to the AST rewrite? It seems like it's introducing a good bit of complexity (e.g. bringing in the junk from `bind`) and am trying to figure out what it gets us.

Reply:
This gets us parity with the existing extended protocol which will allow for some possible code reuse, and avoids adding a new method that uses string formatting which has dubious security properties (hard to prove it doesn't have a SQL injection that we're not aware of).

Reply:
I'll give you that modeling things after the extended protocol is a matter of taste. However, we can trivially ensure that the received value, when parsed as an expression, is an `ast::defs::value::Value::String`.

```rust
let val = match mz_sql_parser::parser::parse_expr(
    display::escape_single_quote_string(val),
) {
    Ok(r @ mz_sql_parser::ast::defs::Value::String(..)) => r,
    _ => bail!("not a string!"),
};
```